ddtrace/datadog/src/exporter.cpp

```diff
@@ -0,0 +1,500 @@
+}
+
+void DdogProfExporter::add_tag(ddog_Vec_Tag &tags, std::string_view key, std::string_view val) {
+  ddog_Vec_Tag_PushResult res = ddog_Vec_Tag_push(&tags, to_slice(key), to_slice(val));
+  if (res.tag == DDOG_VEC_TAG_PUSH_RESULT_ERR) {
+    std::cout << "Error pushing tag '" << key << "'->'" << val << "'" << std::endl;
+    std::cout << "  err: " << ddog_Error_message(&res.err).ptr << std::endl;
+  }
+}
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+bool Uploader::upload(const Profile *profile) {
+  ddog_prof_Profile_SerializeResult result = ddog_prof_Profile_serialize(
+      profile->ddog_profile,
+      nullptr,
+      nullptr);
+  if (result.tag != DDOG_PROF_PROFILE_SERIALIZE_RESULT_OK) {
+    std::cout << "Failure serializing pprof" << std::endl;
+    return false;
+  }
+
+  ddog_prof_EncodedProfile *encoded = &result.ok;
+
+  // Write to disk
+//  ddog_prof_Vec_U8 *buffer = &encoded->buffer;
+//  int fd = open("/tmp/auto.pprof", O_CREAT | O_RDWR, 0600);
+//  int n = -1;
+//  if (1 > (n=write(fd, buffer->ptr, buffer->len)))
+//    std::cout << "Could not write" << std::endl;
+//  else if(buffer->len != n)
+//    std::cout << "Wrote pprof (failed)" << std::endl;
+//  else
+//    std::cout << "Wrote pprof" << std::endl;
+//  close(fd);
+
+  ddog_Timespec start = encoded->start;
+  ddog_Timespec end = encoded->end;
+
+  // Attach file
+  ddog_prof_Exporter_File file[] = {
+    {
+      .name = to_slice("auto.pprof"),
+      .file = {
+        .ptr = encoded->buffer.ptr,
+        .len = encoded->buffer.len,
+      },
+    },
+  };
+
+  ddog_prof_Exporter_Slice_File files = {.ptr = file, .len = 1};
+
+  ddog_prof_Exporter_Request_BuildResult build_res = ddog_prof_Exporter_Request_build(
+      ddog_exporter->ptr,
+      start,
+      end,
+      files,
+      nullptr,
+      nullptr,
+      5000);
+  if (build_res.tag == DDOG_PROF_EXPORTER_REQUEST_BUILD_RESULT_ERR) {
+    std::cout << "Could not build request" << std::endl;
+    std::cout << "  " << ddog_Error_message(&build_res.err).ptr << std::endl;
+    ddog_Error_drop(&build_res.err);
+    return false;
+  }
+
+  ddog_prof_Exporter_Request *req = build_res.ok;
+
+  ddog_prof_Exporter_SendResult res = ddog_prof_Exporter_send(ddog_exporter->ptr, &req, nullptr);
+
+  // Close out request
+
+  if (res.tag == DDOG_PROF_EXPORTER_SEND_RESULT_ERR) {
+    std::cout << "Failed to send result to backend" << std::endl;
+    std::cout << "  url: " <<  url << std::endl;
+    std::cout << "  " << ddog_Error_message(&res.err).ptr << std::endl;
+  }
+
+  // We're done exporting, so reset the profile
+  if (!ddog_prof_Profile_reset(profile->ddog_profile, nullptr)) {
+    std::cout << "Unable to reset!" << std::endl;
+    return false;
+  }
+
+//  ddog_prof_Profile_SerializeResult_drop(result);
+//  ddog_prof_Exporter_Request_drop(req);
+//  ddog_prof_Exporter_SendResult_drop(res);
+
+  return true;
+}
+
+Profile::Profile() {
```

Comment:
Just stating the obvious: We will need some state to know what profile types are active ?

Reply:
Indeed!  I originally had that state, but then took it out to mirror what the Python profiler did (which is to populate *every* available profiling type).  This was primarily so that my overhead figures weren't overblown due to trivial effects.

Lately, there's been some interest in fixing the original Python exporter so that it only populates the pprof entries that it needs to.  I think the correct thing to do is catch the configured profiling types and only provide storage for those.